src/notifications/warning-log.ts

```diff
@@ -32,4 +39,15 @@ export default class WarningLog {
     public copyTo (warningLog: WarningLog): void {
         this.messages.forEach(msg => warningLog.addWarning(msg));
     }
+
+    public static creatAddWarningCallback (messageBus?: MessageBus | object, testRun?: TestRun): (message: string) => Promise<void> {
```

Comment:
It's better to avoid callbacks if we can since they decrease code clarity. Besides, you will hardly need callbacks if you use events.

Instead of trying to integrate MessageBus into WarningLog, I propose to use MessageBus as a main way to handle warnings.

1. WarningLog instances are left only in Task and TestRun to preserve compatibility with the old warning system.
2. Theses instances subscribes to add-warning event and process corresponding warnings.
3. All other code should only emit events to the message bus if it need to emit a warning